codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/endpoints/EndpointResolverGenerator.java

```diff
@@ -0,0 +1,514 @@
+
+        var generator = new ExpressionGenerator(scope, this.fnProvider);
+        var fn = conditionalFunc(condition);
+
+        String conditionName;
+        if (condition.getResult().isPresent()) {
+            var ident = condition.getResult().get();
+            conditionName = "_" + ident.asString();
+
+            // Store the condition result so that it can be referenced in the future by the
+            // result identifier.
+            scope = scope.withIdent(new Reference(ident, SourceLocation.NONE), conditionName);
+        } else {
+            conditionName = nameForExpression(fn);
+        }
+
+        if (fn.type() instanceof Type.Option || isConditionalFnResultOptional(condition, fn)) {
+            return goTemplate("""
+                    if exprVal := $target:W; exprVal != nil {
+                        $condName:L := *exprVal
+                        _ = $condName:L
+                        $next:W
+                    }
+                    """,
+                    MapUtils.of(
+                            "condName", conditionName,
+                            "target", generator.generate(fn),
+                            "next", generateRule(rule, remainingConditions, scope.withMember(fn, conditionName))));
+        }
+
+        if (condition.getResult().isPresent()) {
+            return goTemplate("""
+                    $condName:L := $target:W
+                    _ = $condName:L
+                    $next:W
+                    """,
+                    MapUtils.of(
+                            "condName", conditionName,
+                            "target", generator.generate(fn),
+                            "next", generateRule(rule, remainingConditions, scope.withMember(fn, conditionName))));
+        }
+
+        return goTemplate("""
+                if $target:W {
+                    $next:W
+                }
+                """,
+                MapUtils.of(
+                        "target", generator.generate(fn),
+                        "next", generateRule(rule, remainingConditions, scope)));
+    }
+
+    private static Expression conditionalFunc(Condition condition) {
+        var fn = condition.getFn();
+        if (fn instanceof IsSet) {
+            return ((IsSet) fn).getTarget();
+        }
+        return fn;
+    }
+
+    private String nameForExpression(Expression expr) {
+        conditionNameCounter++;
+        if (expr instanceof Reference) {
+            return nameForRef((Reference) expr);
+        }
+        return String.format("_var_%d", conditionNameCounter);
+    }
+
+    /**
+     * Returns a name for a reference.
+     *
+     * @param ref reference to get name for
+     * @return name
+     */
+    private static String nameForRef(Reference ref) {
+        return "_" + ref.getName();
+    }
+
+    private GoWriter.Writable generateEndpoint(Endpoint endpoint, Scope scope) {
+        return goTemplate("""
+                $endpointType:T{
+                    URI: $url:W,
```

Comment:
**Question**: 

This generates code like this:

```go
URI: func() string {
	var out strings.Builder
	out.WriteString("https://")
	out.WriteString(_Bucket)
	out.WriteString(".op-")
	out.WriteString(_outpostId)
	out.WriteString(".s3-outposts.")
	out.WriteString(_Region)
	out.WriteString(".")
	out.WriteString(_regionPartition.DnsSuffix)
	return out.String()
}(),
```

Is there not any way to generate this cleaner like:

```go
URI: strings.Join(
    []string {
        "https://",
        _Bucket,
        ".op-",
        _outpostId,
        ".s3-outposts.",
        _Region,
        ".",
        _regionPartition.DnsSuffix
    }
)
```

Overall minor point/question.

Reply:
added a todo to reconsider this